### ISSUE TITLE
Question about CVPR2019

### ISSUE THREAD

**Issue by qiulesun** [zhanghang1989/PyTorch-Encoding#206]
1. In Eq. (2) of your CVPR2019 paper (Co-occurrent Features in Semantic Segmentation), one must first compute U_{xc,k} and V_{xc,k}. How do we get them?
2. Could you deliver your supplementary material to my email (qiulesun@163.com)?

**Comment by txfs1926** [comment]
Hi, I also have a question about CFNet.
In the paper, the description of Table 4 mentions that "mIoU on 60 classes w/ background". I wonder if the number of output channels of CFNet is 60. In other words, edit NUM_CLASS to 60, like argued in #78 .
Thanks!

**Comment by qiulesun** [comment]
@txfs1926 May be another answer can help you. See #179 .

**Comment by qiulesun** [comment]
@zhanghang1989 
You set the seed in train.py, i.e., torch.manual_seed(args.seed).
Why not set seed for GPU, i.e., torch.cuda.manual_seed(args.seed)?

**Comment by qiulesun** [comment]
@zhanghang1989 I have noticed that GCU(NeurIPS2018, paper link: https://papers.nips.cc/paper/8135-beyond-grids-learning-graph-representations-for-visual-recognition) creates visualization of the assignment matrix in their method. It is meaningful to illustrate the proposed method.
Do you consider visualizations of the learned assignment weight matrix in EncNet to further straightforwardly detail your method?

**Comment by zhanghang1989** [comment]
> @zhanghang1989 I have noticed that GCU (NeurIPS2018, paper link: https://papers.nips.cc/paper/8135-beyond-grids-learning-graph-representations-for-visual-recognition.pdf) creates visualization of the assignment matrix in their method. It is meaningful to illustrate the proposed method.
> Do you consider visualizations of the learned assignment weight matrix in EncNet to further straightforwardly detail your method?

Thanks @qiulesun for the suggestion. Actually I am not active in research work now.

**Comment by qiulesun** [comment]
@zhanghang1989
I have run your codes more times and alway found that all features fall into one same codeword and other codeworks provide limited represenataton capability.  The learned weights between all features and codeword c_k depend largely on smoth factor s_k in Eq 1 due to the exp() operation in softmax function. Have you evaluated the effec of smoth factors?

**Comment by qiulesun** [comment]
@zhanghang1989  auto GC = tmp.sum(0).sum(0) in https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/lib/cpu/encoding_cpu.cpp#L44 should be auto GC = -tmp.sum(0).sum(0) ?

**Comment by StacyYang** [comment]
> @zhanghang1989 GC = tmp.sum(0).sum(0) in https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/encoding/lib/cpu/encoding_cpu.cpp#L44 should be GC = -tmp.sum(0).sum(0) ?

The weights can learn the negative value automatically  if necessary.

**Comment by qiulesun** [comment]
@StacyYang @zhanghang1989 
In the backword, the computation of  gradC=(-GE.unsqueeze(1) * A.unsqueeze(3)).sum(0).sum(0),
likewise, according to the formulation, dosed the GC is equl to -tmp.sum(0).sum(0) ?

**Comment by qiulesun** [comment]
@zhanghang1989 
I'm very sorry for the inappropriate question.
I deeply want to debug the cuda codes of implementing Encoding module. Could you offer ideas on how to debug them ?

**Comment by qiulesun** [comment]
@zhanghang1989 Could you provide the link to downlond the  model encnet_resnet101_coco pre-trained on COCO datasset ?

**Comment by qiulesun** [comment]
@zhanghang1989 When args.test_val=Ture （https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/experiments/segmentation/test.py#L103), dst (https://github.com/zhanghang1989/PyTorch-Encoding/blob/master/experiments/segmentation/test.py#L103) is the image name rather than the mask. Why is the image name obtained at this time?

**Comment by zhanghang1989** [comment]
Just produce the mask with the same name as the image.
